include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +38,87 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+        {
+            for (uint32_t i = 0u; i < count; ++i)
+            {
+                auto layout = core::smart_refctd_ptr<const IGPUDescriptorSetLayout>(layouts[i]);
+                output[i++] = createDescriptorSet(std::move(layout));
+            }
+        }
+        
+        core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout);
+        bool freeDescriptorSets(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets);
 
-        uint32_t getCapacity() const { return m_maxSets; }
+        inline uint32_t getCapacity() const { return m_maxSets; }
 
     protected:
+        explicit IDescriptorPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, const IDescriptorPool::E_CREATE_FLAGS flags, uint32_t _maxSets, const uint32_t poolSizeCount, const IDescriptorPool::SDescriptorPoolSize* poolSizes) : IBackendObject(std::move(dev)), m_maxSets(_maxSets), m_flags(flags)
+        {
+            std::fill_n(m_maxDescriptorCount, asset::EDT_COUNT, 0u);
+
+            for (auto i = 0; i < poolSizeCount; ++i)
+                m_maxDescriptorCount[poolSizes[i].type] += poolSizes[i].count;
+
+            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+            {
+                if (m_maxDescriptorCount[i] > 0)
+                {
+                    if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+                    {
+                        m_generalAllocatorReservedSpace[i] = std::make_unique<uint8_t[]>(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(1u, m_maxDescriptorCount[i], 1u));
+                        m_generalAllocators[i] = core::GeneralpurposeAddressAllocator<uint32_t>(m_generalAllocatorReservedSpace[i].get(), 0u, 0u, 1u, m_maxDescriptorCount[i], 1u);
+                    }
+                    else
+                    {
+                        m_linearAllocators[i] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[i]);
+                    }
+                }
+            }
+
+            // For (possibly) mutable samplers.
+            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            {
+                m_generalAllocatorReservedSpace[asset::EDT_COUNT] = std::make_unique<uint8_t[]>(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER], 1u));
+                m_generalAllocators[asset::EDT_COUNT] = core::GeneralpurposeAddressAllocator<uint32_t>(m_generalAllocatorReservedSpace[asset::EDT_COUNT].get(), 0u, 0u, 1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER], 1u);
+            }
+            else
+            {
+                m_linearAllocators[asset::EDT_COUNT] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            }
```

Comment:
you could name the union and give it a constructor, then you wouldn't need to repeat yourself, because right now you have an inconsistency

you wont create an allocator if there are no descriptors of a given type, but will do if there are no mutable samplers

Reply:
well instead of having this
```cpp
        uint32_t m_maxDescriptorCount[asset::EDT_COUNT];
        union
        {
            core::LinearAddressAllocator<uint32_t> m_linearAllocators[asset::EDT_COUNT+1];
            core::GeneralpurposeAddressAllocator<uint32_t> m_generalAllocators[asset::EDT_COUNT+1];
        };
        std::unique_ptr<uint8_t[]> m_generalAllocatorReservedSpace[asset::EDT_COUNT+1];
```
you'd have to have this
```cpp
    struct allocator_state_t
    {
        union
        {
            core::LinearAddressAllocator<uint32_t> m_linearAllocator;
            core::GeneralpurposeAddressAllocator<uint32_t> m_generalAllocator;
        };
        std::unique_ptr<uint8_t[]> m_generalAllocatorReservedSpace;
        uint32_t m_maxCount;
   }     
   allocator_state_t m_descriptorAllocators[asset::EDT_COUNT+1];
```

Reply:
A great idea, indeed.

Reply:
Resolved.